test/integration/job/job_test.go

```diff
@@ -1174,6 +1174,242 @@ func TestBackoffLimitPerIndex(t *testing.T) {
+				// happen. We wait 100ms for the sync itself, because we already
+				// checked the metric is incremented so the sync would start
+				// immediately if it was queued.
+				time.Sleep(100 * time.Millisecond)
+				jobObj, err = clientSet.BatchV1().Jobs(jobObj.Namespace).Get(ctx, jobObj.Name, metav1.GetOptions{})
+				if err != nil {
+					t.Fatalf("Error %v when getting the latest job %v", err, klog.KObj(jobObj))
+				}
+				if diff := cmp.Diff(batchv1.JobStatus{}, jobObj.Status); diff != "" {
+					t.Fatalf("Unexpected status (-want/+got): %s", diff)
+				}
+			}
+		})
+	}
+}
+
+// TestManagedByLabel_RecreatedJob verifies that the Job controller skips
+// reconciliation of a job with managed-by label, when this is a recreated job,
+// and there is still a pending sync queued for the previous job.
+// In this scenario we first create a job without managed-by label, and we mark
+// its pod as succeeded. This queues the Job object sync with 1s delay. Then,
+// without waiting for the Job status update we delete and recreate the job under
+// the same name, but with managed-by label. The queued update starts to execute
+// on the new job, but is skipped.
+func TestManagedByLabel_RecreatedJob(t *testing.T) {
+	defer featuregatetesting.SetFeatureGateDuringTest(t, feature.DefaultFeatureGate, features.JobManagedByLabel, true)()
+
+	closeFn, restConfig, clientSet, ns := setup(t, "managed-by-recreate-job")
+	defer closeFn()
+	ctx, cancel := startJobControllerAndWaitForCaches(t, restConfig)
+	defer cancel()
+	resetMetrics()
+
+	baseJob := batchv1.Job{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "custom-job-test",
+			Namespace: ns.Name,
+			Labels:    make(map[string]string),
+		},
+		Spec: batchv1.JobSpec{
+			Completions: ptr.To[int32](1),
+			Parallelism: ptr.To[int32](1),
+			Template: v1.PodTemplateSpec{
+				Spec: v1.PodSpec{
+					Containers: []v1.Container{
+						{
+							Name:  "main-container",
+							Image: "foo",
+						},
+					},
+				},
+			},
+		},
+	}
+	jobObj, err := createJobWithDefaults(ctx, clientSet, ns.Name, &baseJob)
+	if err != nil {
+		t.Fatalf("Error %v when creating the job %q", err, klog.KObj(jobObj))
+	}
+	validateJobPodsStatus(ctx, t, clientSet, jobObj, podsByStatus{
+		Active:      1,
+		Ready:       ptr.To[int32](0),
+		Terminating: ptr.To[int32](0),
+	})
+
+	// Marking the pod as complete queues the job reconciliation
+	if err, _ := setJobPodsPhase(ctx, clientSet, jobObj, v1.PodSucceeded, 1); err != nil {
+		t.Fatalf("Error %v when setting phase %s on the pod of job %v", err, v1.PodSucceeded, klog.KObj(jobObj))
+	}
+
+	jobClient := clientSet.BatchV1().Jobs(jobObj.Namespace)
+	if err = jobClient.Delete(ctx, jobObj.Name, metav1.DeleteOptions{
+		// Use propagationPolicy=background so that we don't need to wait for the job object to be gone.
+		PropagationPolicy: ptr.To(metav1.DeletePropagationBackground),
+	}); err != nil {
+		t.Fatalf("Error %v when deleting the job %v", err, klog.KObj(jobObj))
+	}
+
```

Comment:
this part for testing a regression seems racy. Can we occupy the controller by generating a non trivial number of mock jobs, before queuing our desired change that should be skipped?

Reply:
Possibly, but this is integration test, so we are in charge of what is the scenario. Note that in the line `closeFn, restConfig, clientSet, ns := setup(t, "managed-by-recreate-job")` we spin API server just for this test.

In that setup the test fails consistently when I revert the skip in `syncJob` (the purpose of the test is to prove it is needed there). Also, it is not that racy actually, we need sleep (~1s) before recreating the job so that the test passes. 1s delay is a lot I would say.

Still, Indeed time-based, but it should never fail. If the testing machine is very loaded it may occasionally pass even if the code is broken, but I didn't get this to happen locally. What we expect from the test is that if one day the code is broken we will get notified soon (with failure rate >99% this should be satisfied).

Reply:
Okay let's keep it simpler. It might be only problematic in a regression to know why a small number of cases succeeded. But in the non-broken scenario we are fine.